File: python_producer/producer.py
import pika
import json
import sys

def main():
    # Connect to RabbitMQ
    credentials = pika.PlainCredentials('user', 'password')
    connection = pika.BlockingConnection(
        pika.ConnectionParameters(host='localhost', credentials=credentials))
    channel = connection.channel()

    # Declare a durable queue
    queue_name = 'python_queue'
    channel.queue_declare(queue=queue_name, durable=True)

    # Create a message
    message = {
        'source': 'Python Producer',
        'payload': 'This is the the Python message.'
    }
    body = json.dumps(message)

    # Publish the message
    channel.basic_publish(
        exchange='',
        routing_key=queue_name,
        body=body,
        properties=pika.BasicProperties(
            delivery_mode=2,  # make message persistent
        ))

    print(f" [🚀] Sent {message}")

    # Close the connection
    connection.close()

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print('Interrupted')
        sys.exit(0)
    except Exception as e:
        print(f"An error occurred: {e}")
        sys.exit(1)

File: odoo_producer/producer.js
const amqp = require('amqplib');

// Configuration for the RabbitMQ connection
const rabbitmqHost = 'amqp://user:password@localhost';
const queueName = 'odoo_events';

async function sendMessage() {
  let connection;
  try {
    // 1. Connect to RabbitMQ server
    connection = await amqp.connect(rabbitmqHost);
    console.log('✅ Connected to RabbitMQ');

    // 2. Create a channel
    const channel = await connection.createChannel();
    console.log('✅ Channel created');

    // 3. Assert a queue exists (or create it if it doesn't)
    await channel.assertQueue(queueName, {
      durable: true // The queue will survive a broker restart
    });
    console.log(`✅ Queue '${queueName}' asserted`);

    // 4. Create a sample message representing an Odoo event
    const message = {
      eventId: `evt-${Date.now()}`,
      eventType: 'PRODUCT_UPDATED',
      timestamp: new Date().toISOString(),
      payload: {
        productId: 'PROD-98765',
        name: 'New Product Name',
        price: 199.99
      }
    };
    const messageBuffer = Buffer.from(JSON.stringify(message));

    // 5. Send the message to the queue
    channel.sendToQueue(queueName, messageBuffer, {
      persistent: true // The message will be saved to disk
    });

    console.log(`🚀 Sent message from Odoo to queue '${queueName}':`);
    console.log(JSON.stringify(message, null, 2));

    // 6. Close the channel and connection
    await channel.close();
    await connection.close();
    console.log('✅ Connection closed');

  } catch (error) {
    console.error('❌ Error in Odoo producer:', error);
    if (connection) {
      await connection.close();
      console.log('✅ Connection closed due to error');
    }
    process.exit(1);
  }
}

sendMessage();